crates/forge_services/src/mcp.rs

```
@@ -0,0 +1,336 @@
+            .url
+            .ok_or_else(|| anyhow::anyhow!("URL is required for HTTP server"))?;
+        let transport = rmcp::transport::SseTransport::start(url)
+            .await
+            .map_err(|e| anyhow::anyhow!("Failed to connect server: {e}"))?;
+
+        let client = Self::client_info()
+            .serve(transport)
+            .await
+            .map_err(|e| anyhow::anyhow!("Failed to serve client: {e}"))?;
+
+        let tools = client
+            .list_tools(None)
+            .await
+            .map_err(|e| anyhow::anyhow!("Failed to list tools: {e}"))?;
+        let client = Arc::new(RunnableService::Http(client));
+
+        self.insert_tools(server_name, tools, client.clone())
+            .await
+            .map_err(|e| anyhow::anyhow!("Failed to insert tools: {e}"))?;
+
+        Ok(())
+    }
+    async fn init_mcp(&self, mcp: HashMap<String, McpConfig>) -> anyhow::Result<()> {
+        let http_results: Vec<Option<anyhow::Result<()>>> = futures::future::join_all(
+            mcp.iter()
+                .map(|(server_name, server)| async move {
+                    if self
+                        .servers
+                        .lock()
+                        .map(|v| v.values().any(|v| v.name.eq(server_name)))
+                        .await
+                    {
+                        None
+                    } else if server.url.is_some() {
+                        Some(self.connect_http_server(server_name, server.clone()).await)
+                    } else {
+                        Some(self.connect_stdio_server(server_name, server.clone()).await)
+                    }
+                })
+                // TODO: use flatten function provided by FuturesExt
+                .collect::<Vec<_>>(),
+        )
+        .await;
+
+        for i in http_results.into_iter().flatten() {
+            if let Err(e) = i {
+                tracing::error!("Failed to connect server: {e}");
+            }
+        }
```

Should throw the error.